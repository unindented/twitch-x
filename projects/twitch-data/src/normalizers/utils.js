export function valueOrNull (value) {
  return value || null
}
