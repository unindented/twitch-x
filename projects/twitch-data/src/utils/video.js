import {stringify} from 'qs'
import {USHER_API_URL} from '../constants'

export function getPlaylistUrl ({stream, sig, token}) {
  const query = {
    sig,
    token,
    allow_audio_only: true,
    allow_source: true,
    player: 'twitchweb',
    type: 'any',
    p: Math.floor(Math.random() * 1e7)
  }
  return `${USHER_API_URL}/channel/hls/${stream}.m3u8?${stringify(query)}`
}

export function extractFeedsFromPlaylist (input) {
  const regexp = /#EXT-X-MEDIA:.*NAME="([^"]+)".*\n#EXT-X-STREAM-INF:.*\n(.*)/g
  const result = []

  let match = regexp.exec(input)
  while (match != null) {
    result.push({
      name: match[1],
      url: match[2]
    })
    match = regexp.exec(input)
  }

  return result
}
