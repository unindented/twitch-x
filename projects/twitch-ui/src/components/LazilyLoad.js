import React, {PureComponent} from 'react'
import PropTypes from 'prop-types'
import Spinner from './Spinner'

export default class LazilyLoad extends PureComponent {
  constructor (props) {
    super(props)

    this.state = {
      mod: null
    }
  }

  componentWillMount () {
    this.load(this.props)
  }

  componentWillReceiveProps (nextProps) {
    if (nextProps.load !== this.props.load) {
      this.load(nextProps)
    }
  }

  load (props) {
    this.setState({
      mod: null
    })
    props.load((mod) => {
      this.setState({
        mod: mod.default ? mod.default : mod
      })
    })
  }

  render () {
    return this.state.mod
      ? this.props.children(this.state.mod)
      : <Spinner />
  }
}

LazilyLoad.propTypes = {
  load: PropTypes.func.isRequired,
  children: PropTypes.func.isRequired
}
