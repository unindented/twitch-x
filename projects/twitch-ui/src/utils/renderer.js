import React from 'react'
import renderer from 'react-test-renderer'
import {ThemeProvider} from 'styled-components'
import themes from '../themes'

export function renderWithTheme (node, theme = 'default') {
  return renderer.create(
    <ThemeProvider theme={themes[theme]}>
      {node}
    </ThemeProvider>
  )
}
