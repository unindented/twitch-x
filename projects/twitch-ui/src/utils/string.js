export function supplant (str, obj) {
  return str.replace(/\{\s*([^{}]+?)\s*\}/g, (str, key) => obj[key])
}
