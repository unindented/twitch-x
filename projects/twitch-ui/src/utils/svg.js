export function rectangle ({width, height, fill}) {
  return `<svg viewBox="0 0 ${width} ${height}" xmlns="http://www.w3.org/2000/svg"><rect fill="${fill}" width="100%" height="100%" /></svg>`
}
