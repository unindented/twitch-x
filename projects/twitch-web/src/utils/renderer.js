import ShallowRenderer from 'react-test-renderer/shallow'

export function render (node, theme = 'default') {
  const renderer = new ShallowRenderer()
  renderer.render(node)
  return renderer.getRenderOutput()
}
