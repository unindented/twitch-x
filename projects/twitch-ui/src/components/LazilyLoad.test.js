import React from 'react'
import {renderWithTheme} from '../utils/renderer'
import LazilyLoad from './LazilyLoad'

jest.useFakeTimers()

describe('LazilyLoad', () => {
  let tree

  beforeEach(() => {
    const load = (callback) => {
      setTimeout(callback.bind(null, <div>OHAI</div>), 5000)
    }

    tree = renderWithTheme(
      <LazilyLoad load={load}>
        {(mod) => mod}
      </LazilyLoad>
    )
  })

  it('renders a spinner', () => {
    expect(tree).toMatchSnapshot()
  })

  it('renders the real component once loaded', () => {
    jest.runAllTimers()
    expect(tree).toMatchSnapshot()
  })
})
