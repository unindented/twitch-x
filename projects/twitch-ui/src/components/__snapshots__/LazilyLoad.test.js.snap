// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LazilyLoad renders a spinner 1`] = `
.c0 {
  position: relative;
  -webkit-animation: iECmZH 1s infinite ease-in-out;
  animation: iECmZH 1s infinite ease-in-out;
  border-color: rgba(100,65,164,0.5);
  border-style: solid;
  border-top-color: #6441a4;
  border-width: 10px;
  text-indent: -9999px;
  -webkit-transform: translateZ(0);
  -ms-transform: translateZ(0);
  transform: translateZ(0);
}

.c0,
.c0::after {
  width: 50px;
  height: 50px;
  border-radius: 50%;
}

<div
  className="c0"
>
  Loading...
</div>
`;

exports[`LazilyLoad renders the real component once loaded 1`] = `
<div>
  OHAI
</div>
`;
